app/services/combined_audio_derivative_creator.rb

```diff
@@ -0,0 +1,170 @@
+        original_file.open(rewindable:false) do |input_audio_io|
+          new_temp_file.write input_audio_io.read until input_audio_io.eof?
+        end
+
+        log "Finished downloading #{original_file.metadata['filename']}"
+        result << new_temp_file
+      end
+      result
+    end
+  end
+
+  # A list of arrays; the first item in each is the UUID of each audio member,
+  # while the second is the *starting point* of that audio w/r/t the combined audio.
+  def calculate_start_times()
+    durations = components.map do |f|
+      duration_of_audio_file_in_seconds(f.path)
+    end
+    sum = 0
+    audio_member_ids = audio_members.map(&:id)
+    end_points = durations.map {|i| sum += i}
+    audio_member_ids.zip([0] + end_points)
+  end
+
+  # Generate ffmpeg command to concatenate a set of audio files using
+  # ffmpeg stream mapping.
+  # This is tricky but documented at
+  # https://trac.ffmpeg.org/wiki/Map .
+  # For now, the audio settings are hardwired.
+  # This does not run the command; it just returns
+  # an array that can be passed to @cmd.run .
+  def args_for_ffmpeg(output_file_path)
+    # ffmpeg -y: overwrite the already-existing temp file
+    ffmpeg_command = ['ffmpeg', '-y']
+
+    # Then a list of input files specified with -i
+    input_files = components.map {|x| [ "-i", x.path] }.flatten
+
+    # List the number of audio streams: one per file
+    stream_list = 0.upto(components.count - 1).to_a.map{ |n| "[#{n}:a]"}.join
+
+    # Specify what to do with the audio streams:
+    #
+    # Stream mapping:
+    #   Stream #0:0 (mp3float) -> concat:in0:a0
+    #   Stream #1:0 (mp3float) -> concat:in1:a0
+    #   Stream #2:0 (mp3float) -> concat:in2:a0
+    #   [ and so on]
+    #   concat -> Stream #0:0 (libmp3lame)
+    #
+    # v=0 means there is no video.
+    filtergraph = "concat=n=#{components.count}:v=0:a=1[a]"
+
+    # Finally, some output options:
+    # -map [a] : map just the audio to the output
+    #
+    # -ac 1 : output should be mono
+    #    see https://trac.ffmpeg.org/wiki/AudioChannelManipulation
+    # -b:a 64K: make the output a constant 64k bitrate
+    #    see https://trac.ffmpeg.org/wiki/Encode/MP3
+    output_quality = ["-map", "[a]", "-ac", "1", "-b:a", "64k"]
+
+    # Put it all together and return:
+    ffmpeg_command +
+      input_files +
+      ['-filter_complex'] +
+      ["#{stream_list}#{filtergraph}"] +
+      output_quality +
+      [output_file_path]
+  end
+
+  def audio_member_files
+    audio_members.map { |asset| asset.file }
+  end
+
+  # Filters out non-audio items
+  def audio_members
+    @audio_members ||= begin
+      work.members.order(:position, :id).select do |member|
+        member.stored? && member.file.mime_type.start_with?('audio')
+      end
+    end
+  end
+
+  # If this checksum changes, you need to regenerate the audio
+  def fingerprint
+    @fingerprint ||= begin
+      digests = audio_members.map(&:sha512).compact
```

Comment:
Since start_times includes the UUID pk, should we make sure the UUID pk hasn't changed, by including it in the fingerprint too? Maybe not necessarily, because we don't expect these to ever change, although they ARE technically mutable. But maybe doens't matter.

Reply:
Eh, no harm including UUID pk as well.